lwb/metalang/spt/spt/src/main/java/mb/spt/expectation/RunExpectation.java

```diff
@@ -0,0 +1,139 @@
+        KeyedMessagesBuilder messagesBuilder,
+        TestCase testCase,
+        LanguageUnderTest languageUnderTest,
+        Session languageUnderTestSession, LanguageUnderTestProvider languageUnderTestProvider,
+        ExecContext context,
+        Region sourceRegion
+    ) throws InterruptedException {}
+
+    @Override
+    public KeyedMessages evaluate(
+        TestCase testCase,
+        LanguageUnderTest languageUnderTest,
+        Session languageUnderTestSession,
+        LanguageUnderTestProvider languageUnderTestProvider,
+        ExecContext context,
+        CancelToken cancel
+    ) throws InterruptedException {
+        final ResourceKey file = testCase.testSuiteFile;
+        final KeyedMessagesBuilder messagesBuilder = new KeyedMessagesBuilder();
+        final LanguageInstance languageInstance = languageUnderTest.getLanguageComponent().getLanguageInstance();
+        if(!(languageInstance instanceof TestableAnalysis)) {
+            messagesBuilder.addMessage("Cannot evaluate parse to ATerm expectation because language instance '" + languageInstance + "' does not implement TestableAnalysis", Severity.Error, file, sourceRegion);
+            return messagesBuilder.build(file);
+        }
+        final TestableAnalysis testableAnalysis = (TestableAnalysis)languageInstance;
+        Option<Region> region = selectionReference.map(
+            (sel) -> testCase.testFragment.getSelections().get(sel.selection - 1)
+        );
+        final Result<IStrategoTerm, ?> result = testableAnalysis.testRunStrategy(languageUnderTestSession, testCase.resource, strategy, parseArguments(arguments, testCase), region, testCase.rootDirectoryHint);
+
+        result
+            .ifOkThrowing(
+                (o) -> {
+                    if (expectFailure) {
+                        messagesBuilder.addMessage("Expected strategy to fail, but it succeeded", Severity.Error, file, sourceRegion);
+                    }
+                    checkAst(o, messagesBuilder, testCase, languageUnderTest, languageUnderTestSession, languageUnderTestProvider, context, sourceRegion);
+                }
+            )
+            .ifErr(
+                (e) -> {
+                    if (e instanceof StrategoException) {
+                        if(!expectFailure) {
+                            messagesBuilder.addMessage(
+                                ((StrategoException)e).getMessage(),
+                                e,
+                                Severity.Error,
+                                file,
+                                sourceRegion
+                            );
+                        }
+                    } else {
+                        messagesBuilder.extractMessagesRecursively(e);
+                    }
+                }
+            );
+
+        return messagesBuilder.build();
+    }
+
+    private Option<ListView<RunArgument>> parseArguments(Option<ListView<IStrategoAppl>> arguments, TestCase testCase) {
```

Comment:
Could this be moved to `RunExpectationFromTerm`, as this seems to be converting SPT terms?

Reply:
Perhaps `parseArguments` isn't the best name for this function, and something such as `processArguments` would be more fitting. Unfortunately it can't be moved (entirely) to `RunExpectationFromTerm` because this function converts selection references (an index) to a region, this region is stored as part of the `testCase` passed to `RunExpectation.evaluate`.

Reply:
That makes sense.